### ISSUE TITLE
[Breaking Change Request] [vm/ffi] Disallow empty structs

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#44622]
**Update 2021-01-22**: We will deprecate empty structs as of Dart 2.12 and remove support for them in 2.13.

# Summary

Disallow defining subclasses of `Struct` without fields.

https://github.com/dart-lang/sdk/blob/897b1a2121748a3ee8c7b252b38c6efa6378888d/tests/ffi/vmspecific_static_checks_test.dart#L486

Introduce the `Opaque` native type to represent opaque C types.

```dart
/// [Opaque]'s subtypes represent opaque types in C.
///
/// [Opaque]'s subtypes are not constructible in the Dart code and serve purely
/// as markers in type signatures.
abstract class Opaque extends NativeType {}
```

# Motivation

1. Simpler error messages. Currently we have error messages on using empty structs in nested structs and trampolines. Having these error messages on the empty structs themselves would be easier.
2. Prevent people shooting themselves in the foot by calling `allocate<Utf8>()` and allocating 0 bytes of native memory (https://github.com/dart-lang/sdk/issues/43974, https://github.com/dart-lang/ffi/issues/34).

# Impact

Definitions need to be migrated from `extends Struct` to `extends Opaque`.

Use sites can no longer use `.ref` to pass around 'references' with types such as `Utf8`. Instead, `Pointer<Utf8>` needs to be passed around.

# Migration of package:ffi

Proposed changes:

* https://github.com/dart-lang/ffi/blob/api/opaque/lib/src/utf8.dart
* https://github.com/dart-lang/ffi/blob/api/opaque/lib/src/utf16.dart

# Migration of users of package:ffi

Example migrations:

* https://github.com/dart-lang/ffi/blob/api/opaque/test/utf8_test.dart
* https://dart-review.googlesource.com/c/sdk/+/177710/14/samples/ffi/sqlite/lib/src/database.dart

# Discussion

We could change the `NativeType` hierarchy to the following and reduce the amount of custom checks in the analyzer and common frontend (for `sizeOf`, `Pointer.elementAt` and `Allocator.call`).

* NativeType
  * Opaque
    * Void
    * NativeFunction
    * (user defined types)
  * Sized (only accept these in `sizeOf`, `Pointer.elementAt` (would become an extension method) and `Allocator.call`)
    * NativeInteger
      * ...
    * NativeDouble
    * Struct
        * (user defined types)

**Comment by franklinyow** [comment]
@dcharkes  Please make an announcement to https://groups.google.com/a/dartlang.org/g/announce

cc @Hixie @matanlurey @vsmenon for review and approval.

**Comment by Hixie** [comment]
I have no opinion on this and thus no objection.

**Comment by vsmenon** [comment]
lgtm

**Comment by matanlurey** [comment]
I don't work in this area, no comment.

**Comment by dcharkes** [comment]
> @dcharkes Please make an announcement to https://groups.google.com/a/dartlang.org/g/announce

Announcement sent: https://groups.google.com/a/dartlang.org/g/announce/c/kZOGrc6-rIg

**Comment by simolus3** [comment]
Is there a rough ETA for these changes? Considering that the apis haven't landed yet and that Dart 2.12 was announced for [ "early 2021"](https://medium.com/dartlang/announcing-dart-null-safety-beta-87610fee6730), I'm a bit worried about how little time we might have to migrate.

Is there any way to get the `Opaque` interface into the next beta?

**Comment by franklinyow** [comment]
Approved

**Comment by dcharkes** [comment]
> Is there a rough ETA for these changes?

I've landed the new API in master.

Deprecating the old API will take some time, because this has to roll into Flutter first.

**Comment by dcharkes** [comment]
> Is there a rough ETA for these changes? Considering that the apis haven't landed yet and that Dart 2.12 was announced for [ "early 2021"](https://medium.com/dartlang/announcing-dart-null-safety-beta-87610fee6730), I'm a bit worried about how little time we might have to migrate.
> 
> Is there any way to get the `Opaque` interface into the next beta?

@simolus3 We will deprecate support for empty structs and add `Opaque` in Dart 2.12, and remove support for empty structs in 2.13.

**Comment by simolus3** [comment]
That's relieving to hear, thank you!

**Comment by dcharkes** [comment]
We have landed deprecation hints for this in c3b30df7987d489f1834f115ed6aefc6d66cfd8f.

Turning these hints into errors is tracked in https://github.com/dart-lang/sdk/issues/43974. Closing this issue.
